### DISCUSSION TITLE
Kubernetes service discovery provider based on watch requests

### DISCUSSION THREAD

**Discussion by kick2nick** [ThreeMammals/Ocelot#2168] [Ideas]
## New Feature
New `WatchKube` service discovery provider.
Kube API offers a "watch" request for efficient detection of changes that can be used to keep in sync information about endpoints.

## Motivation for New Feature
Ocelot currently has `Kube` and `PollKube` providers with tradeoffs: fetching endpoint per request with extra hop and pressure on K8s API for `Kube` strategy or possibly outdated endpoints info for `PollKube`. Watch request looks like another option with changes coming more fast than polling and without extra hops/pressure.
I've already implemented custom service discovery provider based on [KubernetesClient](https://github.com/kubernetes-client/csharp) (not [KubeClient](https://github.com/kubernetes-client/csharp)) and solved this for myself, but it would be nice to have it out of the box.

**Comment by kick2nick** [comment]
Found that there were plans for extracting generic polling: https://github.com/ThreeMammals/Ocelot/issues/1708
Is it alive? Any chance that long polling mechanisms such as kube watch api or consul 'blocking queries' will be included?

**Reply by raman-m** [reply]
@ggnaegi  Is it alive?
Why not return to 👉 
- #1910

**Reply by raman-m** [reply]
@kick2nick Welcome to the world of Ocelot!

> Found that there were plans for extracting generic polling: #1708 

From what I gather, this issue aims to standardize the polling services functionality across various discovery providers, rather than introducing a new type of provider.

> Any chance that long polling mechanisms such as kube watch api or consul 'blocking queries' will be included?

I understand [Blocking Queries](https://developer.hashicorp.com/consul/api-docs/features/blocking), but we are not progressing in this direction at the moment. Including new provider types is feasible if you decide to open a PR in the future. Be prepared for an extensive discussion.

Also, my concern is 
> A blocking query is used to wait for a potential change **using long polling.**

Isn't this the same process that `PollKube` employs?

**Reply by kick2nick** [reply]
> Isn't this the same process that PollKube employs?

`PollKube` uses timer with configurable period and makes usual http requests to update list of endpoints.

**Reply by raman-m** [reply]
And what's the problem with regular polling? 😉

**Reply by kick2nick** [reply]
> And what's the problem with regular polling? 

Main problem in [tradeoff ](https://ocelot.readthedocs.io/en/latest/features/kubernetes.html#pollkube-provider). Faster updates put more pressure on k8s api (and ocelot) than it could be with watch requests.

**Comment by raman-m** [comment]
>New `WatchKube` service discovery provider.
**Kube API offers a "watch" request** for efficient detection of changes that can be used to keep in sync information about endpoints.

OK, which one? Any links to docs? Did you mean [Blocking Queries](https://developer.hashicorp.com/consul/api-docs/features/blocking)?

**Reply by kick2nick** [reply]
Blocking queries are about consul.
I'm talking about  [watch requests](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes) that as far as I understand based on chunked http/websocket connection. So basically we need to always keep one open connection per downstream service in order to get changes.

**Reply by raman-m** [reply]
Ups! Sorry, this was my fault! Thanks for correction! 😅 
- [Efficient detection of changes | Kubernetes](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes)

Got it!

**Comment by raman-m** [comment]
> Ocelot currently has `Kube` and `PollKube` providers with tradeoffs: fetching endpoint per request with extra hop and pressure on K8s API for `Kube` strategy or possibly outdated endpoints info for `PollKube`.

While not ideal, this design is functional. If your K8s instance cannot manage the traffic generated by the `Kube` provider, switching to the `PollKube` provider and setting a polling interval is advisable. As for concerns about "possibly outdated endpoints," setting a shorter interval, such as one second, could be a viable solution. Additionally, implementing a retry pattern on the client side is recommended, as some services may be offline due to expiration in the K8s cluster. The total of all retry intervals should match the polling interval to ensure the next poll returns the current list of services. Does that sound reasonable?

On the other hand, we recognize that polling services can lead to bottlenecks, such as having offline services on the list. Perhaps we could consider applying the retry pattern directly on the provider's side. What are your thoughts?

> Watch request looks like another option with changes coming more fast than polling and without extra hops/pressure.

Are you suggesting the use of reactive change logic (aka Publish-Subscribe pattern), such as receiving messages/notifications from a Kubernetes instance? It's a sound concept; however, we must review the documentation for the current library.

**Reply by kick2nick** [reply]
> the retry pattern directly on the provider's side.

Do you mean some sort of reaction on ServiceDiscoveryProvider side when request was sent on endpoint that no longer exist? Sounds reasonable.

> Are you suggesting the use of reactive change logic (aka Publish-Subscribe pattern), such as receiving messages/notifications from a Kubernetes instance?

Exactly, with watch requests mentioned above.
Current library (KubeClient) supports this concept as [Observable](https://github.com/tintoy/dotnet-kube-client/blob/1e08492a6da6ce484ae4b9e5e2ec66a38fe802da/src/KubeClient/ResourceClients/KubeResourceClient.cs#L580)
KubernetesClient supports this concept as [IAsyncEnumerable](https://github.com/kubernetes-client/csharp/blob/541abb00bd1472e86b37863822cce5415613806b/examples/watch/Program.cs#L19)

**Reply by raman-m** [reply]
> Current library (KubeClient) supports this concept as [Observable](https://github.com/tintoy/dotnet-kube-client/blob/1e08492a6da6ce484ae4b9e5e2ec66a38fe802da/src/KubeClient/ResourceClients/KubeResourceClient.cs#L580)

We could develop this idea... Thanks for pointing to the class!

> KubernetesClient supports this concept as [IAsyncEnumerable](https://github.com/kubernetes-client/csharp/blob/541abb00bd1472e86b37863822cce5415613806b/examples/watch/Program.cs#L19)

Sorry, we can't! Maybe later, this year... Firstly, we have to create new repo for the new package which consumes KubernetesClient pack.

**Reply by raman-m** [reply]
> Current library (KubeClient) supports this concept as [Observable](https://github.com/tintoy/dotnet-kube-client/blob/1e08492a6da6ce484ae4b9e5e2ec66a38fe802da/src/KubeClient/ResourceClients/KubeResourceClient.cs#L580)
> KubernetesClient supports this concept as [IAsyncEnumerable](https://github.com/kubernetes-client/csharp/blob/541abb00bd1472e86b37863822cce5415613806b/examples/watch/Program.cs#L19)
> 

Perfect! I have not had the opportunity to explore the KubeClient source code and Kubernetes API in depth, but my curiosity compels me to inquire about the following...
**Do you know which protocol is used to establish a permanent connection between a Kubernetes server and KubeClient as a client? I mean at the physical level.**
HTTP/1.1 is not applicable due to its stateless design. I suppose the possible candidates are HTTP/2/3 or WebSocket. Which one implements the Ping-Pong strategy?

**Reply by kick2nick** [reply]
> Do you know which protocol is used to establish a permanent connection between a Kubernetes server and KubeClient as a client?

Http/1.1. It acts like long polling but without closing connection. When something changes on k8s side it just sends it to KubeClient. Here how it looks like from [docs](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes):
```
GET /api/v1/namespaces/test/pods?watch=1&resourceVersion=10245
---
200 OK
Transfer-Encoding: chunked
Content-Type: application/json

{
  "type": "ADDED",
  "object": {"kind": "Pod", "apiVersion": "v1", "metadata": {"resourceVersion": "10596", ...}, ...}
}
{
  "type": "MODIFIED",
  "object": {"kind": "Pod", "apiVersion": "v1", "metadata": {"resourceVersion": "11020", ...}, ...}
}
...
```

**Comment by raman-m** [comment]
> I've already implemented custom service discovery provider based on [KubernetesClient](https://github.com/kubernetes-client/csharp) (not [KubeClient](https://github.com/kubernetes-client/csharp)) and solved this for myself, but it would be nice to have it out of the box.

It's great, but integration isn't possible at the moment due to the current use of the [KubeClient](https://github.com/kubernetes-client/csharp) package. However, there are plans to transition to the [KubernetesClient](https://github.com/kubernetes-client/csharp) package, as it is more commonly utilized and is the industry standard for C#. Further details can be found in the follow-up ToDo task linked here: https://github.com/ThreeMammals/Ocelot/pull/2111#issuecomment-2274111187 (see the 2nd task).

**Reply by kick2nick** [reply]
`KubernetesClient` was chosen as fast way with documented ability to use `watch` api. But `KubeClient` could also be used for implementing this feature.

**Reply by raman-m** [reply]
Yes, it could! However, my personal capacity is currently exceeded due to a full backlog. Are you planning to develop a proof of concept utilizing [Observable](https://github.com/tintoy/dotnet-kube-client/blob/1e08492a6da6ce484ae4b9e5e2ec66a38fe802da/src/KubeClient/ResourceClients/KubeResourceClient.cs#L580) from the [KubeClient](https://github.com/kubernetes-client/csharp) package (our officially used library)?

**Reply by kick2nick** [reply]
If it would be helpful then no problem, will try to provide proof of concept on this(maybe next) week.
b.t.w. all your links point to [KubernetesClient](https://github.com/kubernetes-client/csharp) and none to tintoy [KubeClient](https://github.com/tintoy/dotnet-kube-client)

**Comment by raman-m** [comment]
### Official pull request for development:
- https://github.com/ThreeMammals/Ocelot/pull/2174
